src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,620 @@
+ *  - its distribution method will be DISTRIBUTE_BY_NONE,
+ *  - its replication model will be ReplicationModel,
+ *  - its replication factor will be set to 1.
+ * Similar to reference tables, it has only 1 placement. In addition to that, that
+ * single placement is only allowed to be on the coordinator.
+ */
+Datum
+create_citus_local_table(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	Oid relationId = PG_GETARG_OID(0);
+
+	/*
+	 * Lock target relation with an exclusive lock - there's no way to make
+	 * sense of this table until we've committed, and we don't want multiple
+	 * backends manipulating this relation.
+	 */
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("could not create citus local table: "
+							   "relation does not exist")));
+	}
+
+	ErrorIfUnsupportedCreateCitusLocalTable(relationId);
+
+	ObjectAddress tableAddress = { 0 };
+	ObjectAddressSet(tableAddress, RelationRelationId, relationId);
+
+	/*
+	 * Ensure dependencies first as we will create shell table on the other nodes
+	 * in the MX case.
+	 */
+	EnsureDependenciesExistOnAllNodes(&tableAddress);
+
+	CreateCitusLocalTable(relationId);
+
+	relation_close(relation, NoLock);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * ErrorIfUnsupportedCreateCitusLocalTable errors out if we cannot create the
+ * citus local table from the relation relationId.
+ */
+static void
+ErrorIfUnsupportedCreateCitusLocalTable(Oid relationId)
+{
+	/* citus local tables can only be created from coordinator for now */
+	EnsureCoordinator();
+
+	EnsureTableOwner(relationId);
+
+	/* we allow creating citus local tables only from relations with RELKIND_RELATION */
+	EnsureRelationKindSupported(relationId);
+
+	if (PartitionTable(relationId) || PartitionedTable(relationId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("citus local tables can not be involved in a "
+							   "partition relationship")));
+	}
+
+	EnsureTableNotDistributed(relationId);
+
+	if (!CoordinatorAddedAsWorkerNode())
+	{
+		const char *relationName = get_rel_name(relationId);
+
+		Assert(relationName != NULL);
+
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg(
+							"cannot create citus local table \"%s\", citus local "
+							"tables can only be created from coordinator node if "
+							"it is added to pg_dist_node", relationName)));
+	}
+}
+
+
+/*
+ * CreateCitusLocalTable is the internal method to create citus local table's
+ * shard and metadata.
+ * Note that this function does not perform any validation on the table as it
+ * is already done in create_citus_local_table.
+ */
+static void
+CreateCitusLocalTable(Oid shellRelationId)
+{
+	/*
+	 * Make sure that existing reference tables have been replicated to all
+	 * the nodes such that we can create foreign keys and joins work
+	 * immediately after creation.
+	 */
+	EnsureReferenceTablesExistOnAllNodes();
```

Comment:
What about moving all the things in `create_citus_local_table` function except the `Oid relationId = PG_GETARG_OID(0);` into the `CreateCitusLocalTable` function. 

This would make `CreateCitusLocalTable` a full internal method that can be called outside of the udf too ?

The structure we follow for other `create_.*_table` udf's is to use the `CreateDistributedTable` function for different kind of citus table types generically. But for citus local tables, it might make more sense to move all the things in `create_citus_local_table` function except the `Oid relationId = PG_GETARG_OID(0);` into the `CreateCitusLocalTable` function.

Reply:
I think we can create wrapper method for that where it does the validation and calls `CreateCitusLocalTable`,  but moving all the checks here without a wrapper method might things hard to understand. Also see https://github.com/citusdata/citus/pull/3852#discussion_r433733376

Reply:
(Assuming you mean "a method that takes a function pointer and executes it" with wrapper method)

However, to be able to write a wrapper method for `Create.*Table` functions, number & type of the arguments of `CreateDistributedTable` and `CreateCitusLocalTable` should be same. But this is not the case.

Did I get it correctly ?

Utilizing "a post validation function" would still be possible if we prefer to push the logic down to the `CreateCitusLocalTable` function.

Reply:
I mean the following (as a prototype),

```c
void CreateCitusLocalTable(relationId) {
 DoCitusLocalTableValidations(relationId)
 CreteCitusLocalTableInternal(relationId) #this is the current CreateCitusLocalTable code
}
```

Reply:
Isn't `ErrorIfUnsupportedCreateCitusLocalTable` = `DoCitusLocalTableValidations` ?